plugins/experimental/system_stats/system_stats.c

```diff
@@ -0,0 +1,255 @@
+/** @file
+
+  @section license License
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+#include "ts/ink_config.h"
+#include "ts/ink_defs.h"
+#include "ts/ts.h"
+
+#include <dirent.h>
+#include <getopt.h>
+#include <inttypes.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#ifdef HAVE_SYS_SYSINFO_H
+#include <sys/sysinfo.h>
+#endif
+
+#include <limits.h>
+
+#define PLUGIN_NAME "system_stats"
+#define DEBUG_TAG PLUGIN_NAME
+
+// Time in MS to grab the system stats
+#define SYSTEM_STATS_TIMEOUT 5000
+
+// Load Average Strings
+#define LOAD_AVG_ONE_MIN "plugin." PLUGIN_NAME ".loadavg.one"
+#define LOAD_AVG_FIVE_MIN "plugin." PLUGIN_NAME ".loadavg.five"
+#define LOAD_AVG_TEN_MIN "plugin." PLUGIN_NAME ".loadavg.ten"
+
+// Process Strings
+#define CURRENT_PROCESSES "plugin." PLUGIN_NAME ".current_processes"
+
+// Base net stats name, full name needs to populated
+// with NET_STATS.infname.RX/TX.standard_net_stats field
+#define NET_STATS "plugin." PLUGIN_NAME ".net."
+
+#define NET_STATS_DIR "/sys/class/net"
+#define STATISTICS_DIR "statistics"
+
+static int
+statAdd(const char *name, TSRecordDataType record_type, TSMutex create_mutex)
+{
+  int stat_id = -1;
+
+  TSMutexLock(create_mutex);
```

Comment:
This looks like it gets called every time you update your stats, for each one. Seems like a lot of mutex overhead. Can we cache the stats?

Reply:
We could, one of the plugins I used for reference was caching them. I think a simpler solution would be to just pull the mutex usage out of the statAdd and just do it on a per-block of stats basis. So in the setnetstat wrap that whole block in the mutex, in the area where we set load/processes wrap that, etc.

Can take a look once we start putting it in to use